### PR TITLE
Support restore/resize from a larger to smaller cluster

### PR DESCRIPTION
**Pull request by oxhead**

### CODE REVIEW

File: scripts/checkpoint.py [hunk 1]
```diff
@@ -107,6 +107,20 @@ def generate_hash(data, prefix='-'):
     return hash_prefix + prefix + data
 
 
+def get_num_slaves(bucket_name, checkpoint_name):
+    # @TODO: seems s3 does not support regex so performance can be an issue
+    print("aws s3 ls s3://{} --human-readable"
```
Comment (vin0110) [inline scripts/checkpoint.py]: @oxhead should this print be here?

File: scripts/checkpoint.py [hunk 2]
```diff
@@ -381,16 +403,29 @@ def service_dfs(op, bucket, checkpoint, regex):
...
+            num_slaves_source = get_num_slaves(bucket, checkpoint)
+            for i in range(num_slaves_source):
+                node_ip = node_list[i % len(node_list)]
+                slave_index = i + 1
+                cmd = generate_cmd(
+                    op, bucket, checkpoint, regex,
+                    slave_index, CheckpointService.service_output)
+                run_cmd(agent, cmd)
+    # only used for in the restore operation
+    num_slaves = num_slaves_source if op == 'restore' else len(node_list)
```
Comment (vin0110) [inline scripts/checkpoint.py]: @oxhead I don't like this. Instead can we set num_slaves in the if stmt on lines 410-425?

**Review by vin0110 — COMMENTED** [review]
@oxhead two comments.

**Comment by oxhead** [comment]
@vin0110 I have revised accordingly.  Thanks.

**Comment by vin0110** [comment]
This was hard for me to follow because the if stmt was eliminated--as it should.

As I said earlier, I am not a fan of using the _if expression_ this way. What is better is an if stmt, ie,
```
if X:
  Y = Z
else:
  Y = W
```
is better than
```
Y = Z if X else W
```

A good use of the if expression is
```
print("{}".format(a.count if a else: 0))
```

**Pull request merged**